homeassistant/bootstrap.py

```diff
@@ -745,113 +744,74 @@ async def _async_resolve_domains_to_setup(
+            all_integrations[dep] = dep_itg
 
-            integrations_to_process.append(itg)
-            manifest_deps.update(itg.dependencies)
-            manifest_deps.update(itg.after_dependencies)
-            if not itg.all_dependencies_resolved:
-                resolve_dependencies_tasks.append(
-                    create_eager_task(
-                        itg.resolve_dependencies(),
-                        name=f"resolve dependencies {domain}",
-                        loop=hass.loop,
-                    )
-                )
-
-        if unseen_deps := manifest_deps - integration_cache.keys():
-            # If there are dependencies, try to preload all
-            # the integrations manifest at once and add them
-            # to the list of requirements we need to install
-            # so we can try to check if they are already installed
-            # in a single call below which avoids each integration
-            # having to wait for the lock to do it individually
-            deps = await loader.async_get_integrations(hass, unseen_deps)
-            for dependant_domain, dependant_itg in deps.items():
-                if isinstance(dependant_itg, loader.Integration):
-                    integration_cache[dependant_domain] = dependant_itg
-                    needed_requirements.update(dependant_itg.requirements)
-
-        if resolve_dependencies_tasks:
-            await asyncio.gather(*resolve_dependencies_tasks)
-
-        for itg in integrations_to_process:
-            try:
-                all_deps = itg.all_dependencies
-            except RuntimeError:
-                # Integration.all_dependencies raises RuntimeError if
-                # dependencies could not be resolved
+    integrations_requirements = {
+        domain: itg.requirements for domain, itg in integrations_to_process.items()
+    }
+    integrations_after_dependencies = (
+        await Integration.resolve_multiple_after_dependencies(
+            hass, integrations_to_process.values(), ignore_exceptions=True
+        )
+    )
+    for deps in integrations_after_dependencies.values():
+        for dep in deps:
+            if dep in integrations_requirements:
                 continue
-            for dep in all_deps:
-                if dep in domains_to_setup:
-                    continue
-                domains_to_setup.add(dep)
-                to_resolve.add(dep)
-
-    _LOGGER.info("Domains to be set up: %s", domains_to_setup)
+            dep_itg = loader.async_get_loaded_integration(hass, dep)
+            integrations_requirements[dep] = dep_itg.requirements
+    all_requirements = set(chain.from_iterable(integrations_requirements.values()))
 
     # Optimistically check if requirements are already installed
     # ahead of setting up the integrations so we can prime the cache
-    # We do not wait for this since its an optimization only
+    # We do not wait for this since it is an optimization only
```

Comment:
A third alternative:
```suggestion
    # We do not wait for this since it's an optimization only
```